src/font_file_lister_gdi.cpp

```diff
@@ -16,284 +16,195 @@
-		k *= c1;
-		k = _rotl(k, r1);
-		k *= c2;
-
-		hash ^= k;
-		hash = _rotl(hash, r2) * m + n;
-	}
-
-	hash ^= len;
-	hash ^= hash >> 16;
-	hash *= 0x85ebca6b;
-	hash ^= hash >> 13;
-	hash *= 0xc2b2ae35;
-	hash ^= hash >> 16;
-
-	return hash;
+GdiFontFileLister::GdiFontFileLister(FontCollectorStatusCallback &)
+: dwrite_factory_sh(nullptr, [](IDWriteFactory* p) { p->Release(); })
+, font_collection_sh(nullptr, [](IDWriteFontCollection* p) { p->Release(); })
+, dc_sh(nullptr, [](HDC dc) { DeleteDC(dc); })
+, gdi_interop_sh(nullptr, [](IDWriteGdiInterop* p) { p->Release(); })
+{
+	IDWriteFactory* dwrite_factory;
+	if (FAILED(DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED, __uuidof(IDWriteFactory), reinterpret_cast<IUnknown**>(&dwrite_factory))))
+		throw agi::EnvironmentError("Failed to initialize the DirectWrite Factory");
+	dwrite_factory_sh = dwrite_factory;
+
+	IDWriteFontCollection* font_collection;
+	if (FAILED(dwrite_factory_sh->GetSystemFontCollection(&font_collection, true)))
+		throw agi::EnvironmentError("Failed to initialize the system font collection");
+	font_collection_sh = font_collection;
+
+	HDC dc = CreateCompatibleDC(nullptr);
+	if (dc == nullptr)
+		throw agi::EnvironmentError("Failed to initialize the HDC");
+	dc_sh = dc;
+
+	IDWriteGdiInterop* gdi_interop;
+	if (FAILED(dwrite_factory_sh->GetGdiInterop(&gdi_interop)))
+		throw agi::EnvironmentError("Failed to initialize the Gdi Interop");
+	gdi_interop_sh = gdi_interop;
 }
 
-std::vector<agi::fs::path> get_installed_fonts() {
-	std::vector<agi::fs::path> files;
+CollectionResult GdiFontFileLister::GetFontPaths(std::string const& facename, int bold, bool italic, std::vector<int> const& characters) {
+	CollectionResult ret;
 
-	wchar_t fdir[MAX_PATH];
-	SHGetFolderPathW(NULL, CSIDL_FONTS, NULL, 0, fdir);
-	agi::fs::path font_dir(fdir);
+	int weight = bold == 0 ? 400 :
+				 bold == 1 ? 700 :
+				 bold;
 
-	// System fonts
-	read_fonts_from_key(HKEY_LOCAL_MACHINE, font_dir, files);
+	// From VSFilter
+	//   - https://sourceforge.net/p/guliverkli2/code/HEAD/tree/src/subtitles/RTS.cpp#l45
+	//   - https://sourceforge.net/p/guliverkli2/code/HEAD/tree/src/subtitles/STS.cpp#l2992
+	LOGFONTW lf{};
+	lf.lfCharSet = DEFAULT_CHARSET; // we should use the one specified in the ass file
+	wcsncpy_s(lf.lfFaceName, LF_FACESIZE, agi::charset::ConvertW(facename).c_str(), _TRUNCATE);
+	lf.lfItalic = italic ? -1 : 0;
+	lf.lfWeight = weight;
+	lf.lfOutPrecision = OUT_TT_PRECIS;
+	lf.lfClipPrecision = CLIP_DEFAULT_PRECIS;
+	lf.lfQuality = ANTIALIASED_QUALITY;
+	lf.lfPitchAndFamily = DEFAULT_PITCH | FF_DONTCARE;
+
+	agi::scoped_holder<HFONT> hfont_sh(CreateFontIndirect(&lf), [](HFONT p) { DeleteObject(p); });
+	if (hfont_sh == nullptr)
+		return ret;
 
-	// User fonts
-	read_fonts_from_key(HKEY_CURRENT_USER, font_dir, files);
+	SelectFont(dc_sh, hfont_sh);
 
-	return files;
-}
+	std::wstring selected_name(LF_FACESIZE - 1, L'\0');
+	if (!GetTextFaceW(dc_sh, LF_FACESIZE, &selected_name[0]))
```

Comment:
Instead of `&selected_name[0]` it's better to use `selected_name.data()` here. Apart from just being more idiomatic this is technically required for correctness since `GetTextFaceW` will implicitly override the null terminator of `selected_name`, which the standard explicitly allows for `selected_name.data() + selected_name.size()` and `selected_name[selected_name.size()]` but technically not for `&selected_name[0] + selected_name.size()` (at least if I read the standard correctly. See also https://stackoverflow.com/a/41310431).

Reply:
I tried it, but I get this error:
```
../src/font_file_lister_gdi.cpp(120): error C2664: 'int GetTextFaceW(HDC,int,LPWSTR)' : impossible de convertir l'argument 3 de 'const _Elem *' en 'LPWSTR'
        with
        [
            _Elem=wchar_t
        ]
```

I'm pretty sure it is because we use [``cpp_std=c++14``](https://github.com/arch1t3cht/Aegisub/blob/cd2e894634bd6fe4038861cfc10491e4ea320123/meson.build#L4). It is only since c++17 or higher that we can use ``selected_name.data()`` in this situation. See https://stackoverflow.com/a/12742517/15835974
I guess that ``&selected_name[0]`` may depend on a undefined behaviour in this situation?

Reply:
Ah, yes. The `data` overload returning a non-const pointer was only added in C++17. I missed that.

However, this is a different question from what standard it'd be legal to use `data` this way in. For that, I was originally looking at cppreference, which listed C++11 for [operator[]](https://en.cppreference.com/w/cpp/string/basic_string/operator_at) and no standard requirement at all (so effectively C++17) for [data](https://en.cppreference.com/w/cpp/string/basic_string/data). But looking at the actual (final drafts of the) standards, it appears that cppreference is wrong and the clause for `operator[]` was indeed added with C++17 and the clause for `data` was added for C++20.

So as far as I can see there's no way to make this technically correct in C++11 without the "pad the string by a byte and resize it afterwards hack". However, seeing as TypesettingTools/Aegisub already switched to C++20 and since this is very likely to still work fine in practice even on earlier standards, I'd be fine with keeping this as it is and adding a comment like `FIXME: This will override the string's terminator, which is not technically correct. After switching to C++20 this should use .data().`

Reply:
Addressed in commit 44b81fdfae59a9096c6dcab94cf38d6d38e1052d